source/distributed/tria.cc

```diff
@@ -3775,6 +3776,187 @@ namespace parallel
+              }
+
+          dealii::internal::p4est::functions<dim>::
+          quadrant_childrenv (&p4est_cell, p4est_child);
+
+          // mark first child with CELL_REFINE and the remaining children with CELL_INVALID,
+          // but associate them with all with the parent cell
+          // unpack algorithm shall be called only on CELL_REFINE flagged quadrant
+          for (unsigned int i=0; i<GeometryInfo<dim>::max_children_per_cell; ++i)
+            {
+              int child_idx = sc_array_bsearch(const_cast<sc_array_t *>(&tree.quadrants),
+                                               &p4est_child[i],
+                                               dealii::internal::p4est::functions<dim>::quadrant_compare);
+
+              auto cell_status = (i==0) ? Triangulation<dim,spacedim>::CELL_REFINE
+                                 : Triangulation<dim,spacedim>::CELL_INVALID;
+
+              store_single_quadrant_cell_relation(tree,
+                                                  child_idx,
+                                                  dealii_cell,
+                                                  cell_status);
+            }
+        }
+      else // !p4est_has_children && dealii_cell->has_children())
+        {
+          // its children got coarsened into this cell in p4est, but the dealii_cell
+          // still has its children
+          store_single_quadrant_cell_relation(tree,
+                                              idx,
+                                              dealii_cell,
+                                              Triangulation<dim,spacedim>::CELL_COARSEN);
+        }
+    }
+
+
+
+    template <int dim, int spacedim>
+    inline void
+    Triangulation<dim, spacedim>::
+    store_single_quadrant_cell_relation (const typename dealii::internal::p4est::types<dim>::tree &tree,
+                                         const unsigned int idx,
+                                         const typename Triangulation<dim,spacedim>::cell_iterator &dealii_cell,
+                                         const typename Triangulation<dim,spacedim>::CellStatus status)
+    {
+      const unsigned int local_quadrant_index = tree.quadrants_offset + idx;
+
+      auto q = static_cast<typename dealii::internal::p4est::types<dim>::quadrant *> (
+                 sc_array_index (const_cast<sc_array_t *>(&tree.quadrants), idx)
+               );
+      // TODO CHECK: Do we need the quadrant pointer at all?
+      //             Or is its position in the sc_array sufficient?
+      //             Store pairs (dealii_cell, cell_status) in order of
+      //             appearance of its corresponding quadrant?
+      //             (Then storing the corresponding tree would be interesting?)
+
+      auto tuple = std::make_tuple(q,
+                                   status,
+                                   dealii_cell);
+
+      local_quadrant_cell_relations[local_quadrant_index] = tuple;
```

Comment:
this is a memory error: you only `reserve`d the array, but you never `resize`d it! So it will still have size zero. You need to resize it correctly, but even then it would be useful to add something like
```
  AssertDimension (local_quadrant_index, local_quadrant_cell_relations.size());
```